src/mpi/init/init_async.c

```diff
@@ -59,22 +78,170 @@ static void progress_fn(void *data)
+        if (have_cliques) {
+            fprintf(stderr,
+                    "Setting affinity for progress threads cannot work correctly with MPIR_CVAR_NUM_CLIQUES or MPIR_CVAR_ODD_EVEN_CLIQUES.\n");
+        }
+
+        global_rank = MPIR_Process.rank;
+        local_rank =
+            (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->node_comm->rank : 0;
+        if (have_cliques) {
+            /* If local cliques > 1, using local_size from node_comm will have conflict on thread idx.
+             * In multiple nodes case, this would cost extra memory for allocating thread affinity on every
+             * node, but it is okay to solve progress thread oversubscription. */
+            local_size = MPIR_Process.comm_world->local_size;
+        } else {
+            local_size =
+                (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->
+                node_comm->local_size : 1;
+        }
+
+        async_threads_per_node = local_size;
+        thread_affinity = (int *) MPL_malloc(async_threads_per_node * sizeof(int), MPL_MEM_OTHER);
+
+        MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                        (MPL_DBG_FDEST,
+                         " global_rank %d, local_rank %d, local_size %d, async_threads_per_node %d",
+                         global_rank, local_rank, local_size, async_threads_per_node));
+
+        mpi_errno = MPIDI_parse_progress_thread_affinity(thread_affinity, async_threads_per_node);
+        MPIR_ERR_CHECK(mpi_errno);
+
+        if (MPIR_Process.rank == 0) {
+            int th_idx;
+            for (th_idx = 0; th_idx < async_threads_per_node; th_idx++) {
+                MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                                (MPL_DBG_FDEST, "affinity: thread %d, processor %d",
+                                 th_idx, thread_affinity[th_idx]));
+            }
+        }
+        *p_thread_affinity = thread_affinity;
+        if (have_cliques) {
+            /* In this case, procs on one physical node are partitioned into different virtual nodes,
+             * global_rank should be used to avoid binding progress threads from different ranks to the same core. */
+            *affinity_idx = global_rank;
+        } else {
+            *affinity_idx = local_rank;
+        }
+
+    }
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+#endif
+
 /* called inside MPID_Init_async_thread to provide device override */
 int MPIR_Init_async_thread(void)
 {
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno = MPI_SUCCESS, thr_err;
+    int *thread_affinity = NULL, affinity_idx;
+    bool apply_affinity;
     MPIR_FUNC_TERSE_STATE_DECL(MPID_STATE_MPIR_INIT_ASYNC_THREAD);
 
     MPIR_FUNC_TERSE_ENTER(MPID_STATE_MPIR_INIT_ASYNC_THREAD);
 
+#ifdef DO_ASYNC_THREAD_AFFINITY
+    mpi_errno = get_thread_affinity(&apply_affinity, &thread_affinity, &affinity_idx);
+    MPIR_ERR_CHECK(mpi_errno);
```

Comment:
This can be move and group together with at line 235, right?

Reply:
With the current structure, async threads are not created if the parsing fails. So, I kept it as it is. Would that be a good reason to not move it?

Reply:
Okay, I guess that's fine.